File: input_source.m
#import <Carbon/Carbon.h>

static const char kInstallPath[] = "/Library/Input Methods/Squirrel.app";

static const CFStringRef kHansInputModeID =
    CFSTR("im.rime.inputmethod.Squirrel.Hans");
static const CFStringRef kHantInputModeID =
    CFSTR("im.rime.inputmethod.Squirrel.Hant");
static const CFStringRef kCantInputModeID =
    CFSTR("im.rime.inputmethod.Squirrel.Cant");

typedef NS_OPTIONS(int, RimeInputMode) {
  DEFAULT_INPUT_MODE = 1 << 0,
  HANS_INPUT_MODE = 1 << 0,
  HANT_INPUT_MODE = 1 << 1,
  CANT_INPUT_MODE = 1 << 2
};

RimeInputMode GetEnabledInputModes(void);

void RegisterInputSource(void) {
  if (GetEnabledInputModes()) {  // Already registered
    return;
  }
  CFURLRef installPathURL = CFURLCreateFromFileSystemRepresentation(
      NULL, (UInt8*)kInstallPath, (CFIndex)strlen(kInstallPath), false);
  if (installPathURL) {
    TISRegisterInputSource((CFURLRef)CFAutorelease(installPathURL));
    NSLog(@"Registered input source from %s", kInstallPath);
  }
}

void EnableInputSource(void) {
  RimeInputMode input_modes_enabled = GetEnabledInputModes();
  if (input_modes_enabled != 0) {
    // keep user's manually enabled input modes
    return;
  }
  RimeInputMode input_modes_to_enable = 0;
  CFArrayRef localizations = CFArrayCreate(
      NULL, (CFTypeRef[]){CFSTR("zh-Hans"), CFSTR("zh-Hant"), CFSTR("zh-HK")},
      3, &kCFTypeArrayCallBacks);
  CFArrayRef preferred =
      CFBundleCopyLocalizationsForPreferences(localizations, NULL);
  if (CFArrayGetCount(preferred) > 0) {
    CFStringRef language = (CFStringRef)CFArrayGetValueAtIndex(preferred, 0);
    if (!CFStringCompare(language, CFSTR("zh-Hans"), 0)) {
      input_modes_to_enable |= HANS_INPUT_MODE;
    } else if (!CFStringCompare(language, CFSTR("zh-Hant"), 0)) {
      input_modes_to_enable |= HANT_INPUT_MODE;
    } else if (!CFStringCompare(language, CFSTR("zh-HK"), 0)) {
      input_modes_to_enable |= CANT_INPUT_MODE;
    }
  } else {
    input_modes_to_enable = HANS_INPUT_MODE;
  }
  CFRelease(preferred);
  CFDictionaryRef property = CFDictionaryCreate(
      NULL, (CFTypeRef[]){kTISPropertyBundleID},
      (CFTypeRef[]){CFBundleGetIdentifier(CFBundleGetMainBundle())}, 1,
      &kCFTypeDictionaryKeyCallBacks, &kCFTypeDictionaryValueCallBacks);
  CFArrayRef sourceList = TISCreateInputSourceList(property, true);
  for (CFIndex i = 0; i < CFArrayGetCount(sourceList); ++i) {
    TISInputSourceRef inputSource =
        (TISInputSourceRef)CFArrayGetValueAtIndex(sourceList, i);
    CFStringRef sourceID = (CFStringRef)TISGetInputSourceProperty(
        inputSource, kTISPropertyInputSourceID);
    // NSLog(@"Examining input source: %@", sourceID);
    if ((!CFStringCompare(sourceID, kHansInputModeID, 0) &&
         (input_modes_to_enable & HANS_INPUT_MODE)) ||
        (!CFStringCompare(sourceID, kHantInputModeID, 0) &&
         (input_modes_to_enable & HANT_INPUT_MODE)) ||
        (!CFStringCompare(sourceID, kCantInputModeID, 0) &&
         (input_modes_to_enable & CANT_INPUT_MODE))) {
      CFBooleanRef isEnabled = (CFBooleanRef)TISGetInputSourceProperty(
          inputSource, kTISPropertyInputSourceIsEnabled);
      if (!CFBooleanGetValue(isEnabled)) {
        OSStatus enableError = TISEnableInputSource(inputSource);
        if (enableError) {
          NSLog(@"Failed to enable input source: %@ (%@)", sourceID,
                [NSError errorWithDomain:NSOSStatusErrorDomain
                                    code:enableError
                                userInfo:nil]);
        } else {
          NSLog(@"Enabled input source: %@", sourceID);
        }
      }
    }
  }
  CFRelease(sourceList);
}

void SelectInputSource(void) {
  RimeInputMode enabled_input_modes = GetEnabledInputModes();
  RimeInputMode input_mode_to_select = 0;
  CFArrayRef localizations = CFArrayCreate(
      NULL, (CFTypeRef[]){CFSTR("zh-Hans"), CFSTR("zh-Hant"), CFSTR("zh-HK")},
      3, &kCFTypeArrayCallBacks);
  CFArrayRef preferred =
      CFBundleCopyLocalizationsForPreferences(localizations, NULL);
  for (CFIndex i = 0; i < CFArrayGetCount(preferred); ++i) {
    CFStringRef language = (CFStringRef)CFArrayGetValueAtIndex(preferred, i);
    if (!CFStringCompare(language, CFSTR("zh-Hans"), 0) &&
        (enabled_input_modes & HANS_INPUT_MODE)) {
      input_mode_to_select = HANS_INPUT_MODE;
      break;
    }
    if (!CFStringCompare(language, CFSTR("zh-Hant"), 0) &&
        (enabled_input_modes & HANT_INPUT_MODE)) {
      input_mode_to_select = HANT_INPUT_MODE;
      break;
    }
    if (!CFStringCompare(language, CFSTR("zh-HK"), 0) &&
        (enabled_input_modes & CANT_INPUT_MODE)) {
      input_mode_to_select = CANT_INPUT_MODE;
      break;
    }
  }
  CFRelease(preferred);
  if (input_mode_to_select == 0) {
    NSLog(@"No enabled input sources.");
    return;
  }
  CFDictionaryRef property = CFDictionaryCreate(
      NULL, (CFTypeRef[]){kTISPropertyBundleID},
      (CFTypeRef[]){CFBundleGetIdentifier(CFBundleGetMainBundle())}, 1,
      &kCFTypeDictionaryKeyCallBacks, &kCFTypeDictionaryValueCallBacks);
  CFArrayRef sourceList = TISCreateInputSourceList(property, false);
  for (CFIndex i = 0; i < CFArrayGetCount(sourceList); ++i) {
    TISInputSourceRef inputSource =
        (TISInputSourceRef)CFArrayGetValueAtIndex(sourceList, i);
    CFStringRef sourceID = (CFStringRef)TISGetInputSourceProperty(
        inputSource, kTISPropertyInputSourceID);
    // NSLog(@"Examining input source: %@", sourceID);
    if ((!CFStringCompare(sourceID, kHansInputModeID, 0) &&
         ((input_mode_to_select & HANS_INPUT_MODE) != 0)) ||
        (!CFStringCompare(sourceID, kHantInputModeID, 0) &&
         ((input_mode_to_select & HANT_INPUT_MODE) != 0)) ||
        (!CFStringCompare(sourceID, kCantInputModeID, 0) &&
         ((input_mode_to_select & CANT_INPUT_MODE) != 0))) {
      // select the first enabled input mode in Squirrel.
      CFBooleanRef isSelectable = (CFBooleanRef)TISGetInputSourceProperty(
          inputSource, kTISPropertyInputSourceIsSelectCapable);
      CFBooleanRef isSelected = (CFBooleanRef)TISGetInputSourceProperty(
          inputSource, kTISPropertyInputSourceIsSelected);
      if (!CFBooleanGetValue(isSelected) && CFBooleanGetValue(isSelectable)) {
        OSStatus selectError = TISSelectInputSource(inputSource);
        if (selectError) {
          NSLog(@"Failed to select input source: %@ (%@)", sourceID,
                [NSError errorWithDomain:NSOSStatusErrorDomain
                                    code:selectError
                                userInfo:nil]);
        } else {
          NSLog(@"Selected input source: %@", sourceID);
          break;
        }
      }
    }
  }
  CFRelease(sourceList);
}

void DisableInputSource(void) {
  CFDictionaryRef property = CFDictionaryCreate(
      NULL, (CFTypeRef[]){kTISPropertyBundleID},
      (CFTypeRef[]){CFBundleGetIdentifier(CFBundleGetMainBundle())}, 1,
      &kCFTypeDictionaryKeyCallBacks, &kCFTypeDictionaryValueCallBacks);
  CFArrayRef sourceList = TISCreateInputSourceList(property, false);
  for (CFIndex i = CFArrayGetCount(sourceList); i > 0; --i) {
    TISInputSourceRef inputSource =
        (TISInputSourceRef)CFArrayGetValueAtIndex(sourceList, i - 1);
    CFStringRef sourceID = (CFStringRef)TISGetInputSourceProperty(
        inputSource, kTISPropertyInputSourceID);
    // NSLog(@"Examining input source: %@", sourceID);
    if (!CFStringCompare(sourceID, kHansInputModeID, 0) ||
        !CFStringCompare(sourceID, kHantInputModeID, 0) ||
        !CFStringCompare(sourceID, kCantInputModeID, 0)) {
      OSStatus disableError = TISDisableInputSource(inputSource);
      if (disableError) {
        NSLog(@"Failed to disable input source: %@ (%@)", sourceID,
              [NSError errorWithDomain:NSOSStatusErrorDomain
                                  code:disableError
                              userInfo:nil]);
      } else {
        NSLog(@"Disabled input source: %@", sourceID);
      }
    }
  }
  CFRelease(sourceList);
}

RimeInputMode GetEnabledInputModes(void) {
  RimeInputMode input_modes = 0;
  CFDictionaryRef property = CFDictionaryCreate(
      NULL, (CFTypeRef[]){kTISPropertyBundleID},
      (CFTypeRef[]){CFBundleGetIdentifier(CFBundleGetMainBundle())}, 1,
      &kCFTypeDictionaryKeyCallBacks, &kCFTypeDictionaryValueCallBacks);
  CFArrayRef sourceList = TISCreateInputSourceList(property, false);
  for (CFIndex i = 0; i < CFArrayGetCount(sourceList); ++i) {
    TISInputSourceRef inputSource =
        (TISInputSourceRef)CFArrayGetValueAtIndex(sourceList, i);
    CFStringRef sourceID = (CFStringRef)TISGetInputSourceProperty(
        inputSource, kTISPropertyInputSourceID);
    // NSLog(@"Examining input source: %@", sourceID);
    if (!CFStringCompare(sourceID, kHansInputModeID, 0) ||
        !CFStringCompare(sourceID, kHantInputModeID, 0) ||
        !CFStringCompare(sourceID, kCantInputModeID, 0)) {
      if (!CFStringCompare(sourceID, kHansInputModeID, 0)) {
        input_modes |= HANS_INPUT_MODE;
      } else if (!CFStringCompare(sourceID, kHantInputModeID, 0)) {
        input_modes |= HANT_INPUT_MODE;
      } else if (!CFStringCompare(sourceID, kCantInputModeID, 0)) {
        input_modes |= CANT_INPUT_MODE;
      }
    }
  }
  CFRelease(sourceList);
  return input_modes;
}

File: input_source.swift
import Carbon


struct RimeInputMode: OptionSet {
  let rawValue: CInt
  static let DEFAULT = RimeInputMode(rawValue: 1 << 0)
  static let HANS = RimeInputMode(rawValue: 1 << 0)
  static let HANT = RimeInputMode(rawValue: 1 << 1)
  static let CANT = RimeInputMode(rawValue: 1 << 3)
}


let kInstallPath: String = "/Library/Input Methods/Squirrel.app"

let kHansInputModeID = "im.rime.inputmethod.Squirrel.Hans" as CFString
let kHantInputModeID = "im.rime.inputmethod.Squirrel.Hant" as CFString
let kCantInputModeID = "im.rime.inputmethod.Squirrel.Cant" as CFString

func RegisterInputSource() {
  if (!GetEnabledInputModes().isEmpty) {
    // Already registered
    return
  }
  if let installPathURL = CFURLCreateFromFileSystemRepresentation(
    kCFAllocatorDefault, Array(kInstallPath.utf8), CFIndex(strlen(kInstallPath)), false) {
    TISRegisterInputSource(installPathURL)
    NSLog("Registered input source from %@", kInstallPath)
  }
}

func EnableInputSource() {
  if (!GetEnabledInputModes().isEmpty) {
    // keep user's manually enabled input modes
    return;
  }
  var input_modes_to_enable = RimeInputMode()
  let preferred: CFArray = CFBundleCopyLocalizationsForPreferences(
    ["zh-Hans", "zh-Hant", "zh-HK"] as CFArray, nil)
  if (CFArrayGetCount(preferred) > 0) {
    let language: CFString = CFArrayGetValueAtIndex(preferred, 0) as! CFString
    if (CFStringCompare(language, "zh-Hans" as CFString,
                        [.compareCaseInsensitive]) == .compareEqualTo) {
      input_modes_to_enable.insert(.HANS)
    } else if (CFStringCompare(language, "zh-Hant" as CFString,
                               [.compareCaseInsensitive]) == .compareEqualTo) {
      input_modes_to_enable.insert(.HANT)
    } else if (CFStringCompare(language, "zh-HK" as CFString,
                               [.compareCaseInsensitive]) == .compareEqualTo) {
      input_modes_to_enable.insert(.CANT)
    }
  } else {
    input_modes_to_enable = .HANS
  }
  let property = [kTISPropertyBundleID:
                    CFBundleGetIdentifier(CFBundleGetMainBundle())] as CFDictionary
  let sourceList = TISCreateInputSourceList(property, true) as! CFArray
  for i in 0..<CFArrayGetCount(sourceList) {
    let inputSource = CFArrayGetValueAtIndex(sourceList, i) as! TISInputSource
    let sourceID = TISGetInputSourceProperty(
      inputSource, kTISPropertyInputSourceID) as! CFString
    // NSLog(@"Examining input source: %@", sourceID);
    if ((CFStringCompare(sourceID, kHansInputModeID, []) == .compareEqualTo &&
         input_modes_to_enable.contains(.HANS)) ||
        (CFStringCompare(sourceID, kHantInputModeID, []) == .compareEqualTo &&
         input_modes_to_enable.contains(.HANT)) ||
        (CFStringCompare(sourceID, kCantInputModeID, []) == .compareEqualTo &&
         input_modes_to_enable.contains(.CANT))) {
      let isEnabled = TISGetInputSourceProperty(
        inputSource, kTISPropertyInputSourceIsEnabled) as! CFBoolean
      if (!CFBooleanGetValue(isEnabled)) {
        let enableError: OSStatus = TISEnableInputSource(inputSource)
        if (enableError != 0) {
          NSLog("Failed to enable input source: %@ (%@)",
                [sourceID, NSError(domain: NSOSStatusErrorDomain,
                                   code: Int(enableError), userInfo: nil)])
        } else {
          NSLog("Enabled input source: %@", [sourceID])
        }
      }
    }
  }
}

func SelectInputSource() {
  let enabled_input_modes: RimeInputMode = GetEnabledInputModes();
  var input_mode_to_select = RimeInputMode();
  let preferred: CFArray = CFBundleCopyLocalizationsForPreferences(
    ["zh-Hans", "zh-Hant", "zh-HK"] as CFArray, nil)
  for i in 0..<CFArrayGetCount(preferred) {
    let language = CFArrayGetValueAtIndex(preferred, i) as! CFString
    if (CFStringCompare(language, "zh-Hans" as CFString,
                        [.compareCaseInsensitive]) == .compareEqualTo &&
        enabled_input_modes.contains(.HANS)) {
      input_mode_to_select.update(with: .HANS)
      break
    }
    if (CFStringCompare(language, "zh-Hant" as CFString,
                        [.compareCaseInsensitive]) == .compareEqualTo &&
        enabled_input_modes.contains(.HANT)) {
      input_mode_to_select.update(with: .HANT)
      break
    }
    if (CFStringCompare(language, "zh-HK" as CFString,
                        [.compareCaseInsensitive]) == .compareEqualTo &&
        enabled_input_modes.contains(.CANT)) {
      input_mode_to_select.update(with: .CANT)
      break
    }
  }
  if (input_mode_to_select.isEmpty) {
    NSLog("No enabled input sources.")
    return
  }
  let property = [kTISPropertyBundleID:
                    CFBundleGetIdentifier(CFBundleGetMainBundle())] as CFDictionary
  let sourceList = TISCreateInputSourceList(property, false) as! CFArray
  for i in 0..<CFArrayGetCount(sourceList) {
    let inputSource = CFArrayGetValueAtIndex(sourceList, i) as! TISInputSource
    let sourceID = TISGetInputSourceProperty(
      inputSource, kTISPropertyInputSourceID) as! CFString
    // NSLog(@"Examining input source: %@", sourceID);
    if ((CFStringCompare(sourceID, kHansInputModeID, []) == .compareEqualTo &&
         input_mode_to_select.contains(.HANS)) ||
        (CFStringCompare(sourceID, kHantInputModeID, []) == .compareEqualTo &&
         input_mode_to_select.contains(.HANT)) ||
        (CFStringCompare(sourceID, kCantInputModeID, []) == .compareEqualTo &&
         input_mode_to_select.contains(.CANT))) {
      // select the first enabled input mode in Squirrel.
      let isSelectable = TISGetInputSourceProperty(
        inputSource, kTISPropertyInputSourceIsSelectCapable) as! CFBoolean
      let isSelected = TISGetInputSourceProperty(
        inputSource, kTISPropertyInputSourceIsSelected) as! CFBoolean
      if (!CFBooleanGetValue(isSelected) && CFBooleanGetValue(isSelectable)) {
        let selectError: OSStatus = TISSelectInputSource(inputSource)
        if (selectError != 0) {
          NSLog("Failed to select input source: %@ (%@)",
                [sourceID, NSError(domain: NSOSStatusErrorDomain, code: Int(selectError))])
        } else {
          NSLog("Selected input source: %@", [sourceID])
          break
        }
      }
    }
  }
}

func DisableInputSource() {
  let property = [kTISPropertyBundleID:
                    CFBundleGetIdentifier(CFBundleGetMainBundle())] as CFDictionary
  let sourceList = TISCreateInputSourceList(property, false) as! CFArray
  for i in (0..<CFArrayGetCount(sourceList)).reversed() {
    let inputSource = CFArrayGetValueAtIndex(sourceList, i) as! TISInputSource
    let sourceID = TISGetInputSourceProperty(
      inputSource, kTISPropertyInputSourceID) as! CFString
    // NSLog(@"Examining input source: %@", sourceID);
    if (CFStringCompare(sourceID, kHansInputModeID, []) == .compareEqualTo ||
        CFStringCompare(sourceID, kHantInputModeID, []) == .compareEqualTo ||
        CFStringCompare(sourceID, kCantInputModeID, []) == .compareEqualTo) {
      let disableError: OSStatus = TISDisableInputSource(inputSource)
      if (disableError != 0) {
        NSLog("Failed to disable input source: %@ (%@)",
              [sourceID, NSError(domain: NSOSStatusErrorDomain, code: Int(disableError))])
      } else {
        NSLog("Disabled input source: %@", [sourceID])
      }
    }
  }
}

func GetEnabledInputModes() -> RimeInputMode {
  var input_modes = RimeInputMode()
  let property = [kTISPropertyBundleID:
                    CFBundleGetIdentifier(CFBundleGetMainBundle())] as CFDictionary
  let sourceList = TISCreateInputSourceList(property, false) as! CFArray
  for i in 0..<CFArrayGetCount(sourceList) {
    let inputSource = CFArrayGetValueAtIndex(sourceList, i) as! TISInputSource
    let sourceID = TISGetInputSourceProperty(
      inputSource, kTISPropertyInputSourceID) as! CFString
    // NSLog(@"Examining input source: %@", sourceID);
    if (CFStringCompare(sourceID, kHansInputModeID, []) == .compareEqualTo) {
      input_modes.insert(.HANS)
    } else if (CFStringCompare(sourceID, kHantInputModeID, []) == .compareEqualTo) {
      input_modes.insert(.HANT)
    } else if (CFStringCompare(sourceID, kCantInputModeID, []) == .compareEqualTo) {
      input_modes.insert(.CANT)
    }
  }
  return input_modes
}
